`Exact` does not work as expected with type involving `Date` and union

In the example below, the cases `t7` and `t8` cause a TS error.

```ts
import type { Exact } from "type-fest";

interface A {
	a: string;
	b?: Date | null;
}

function testArg<T extends Exact<A, T>>(a: T) {
	return a;
}

const t1 = { a: "a" };
testArg(t1); // OK - Type of arg correct, no TS error

const t2 = { a: "a", b: new Date() };
testArg(t2); // OK - Type of arg correct, no TS error

const t3 = { a: "a", b: null };
testArg(t3); // OK - Type of arg correct, no TS error

const t4 = { a: "a", b: undefined };
testArg(t4); // OK - Type of arg correct, no TS error

const t5 = { a: "a", b: new Date() as Date | null | undefined };
testArg(t5); // OK - Type of arg correct, no TS error

const t6 = { a: "a", b: null as null | undefined };
testArg(t6); // OK - Type of arg correct, no TS error

const t7 = { a: "a", b: new Date() as Date | undefined };
testArg(t7); // Bug - Type of arg correct, but TS error

const t8 = { a: "a", b: new Date() as Date | null };
testArg(t8); // Bug - Type of arg correct, but TS error
```

Error at `testArg(t7)`:
```
error TS2345: Argument of type '{ a: string; b: Date | undefined; }' is not assignable to parameter of type 'ExactObject<A, { a: string; b: Date | undefined; }>'.
  Type '{ a: string; b: Date | undefined; }' is not assignable to type '{ a: string; b?: ExactObject<Date, Date | undefined> | null | undefined; }'.
    Types of property 'b' are incompatible.
      Type 'Date | undefined' is not assignable to type 'ExactObject<Date, Date | undefined> | null | undefined'.
        Type 'Date' is not assignable to type 'ExactObject<Date, Date | undefined>'.
          Type 'Date' is not assignable to type '{ toString: ExactObject<() => string, never>; toDateString: ExactObject<() => string, never>; toTimeString: ExactObject<() => string, never>; ... 41 more ...; [Symbol.toPrimitive]: ExactObject<...>; }'.
            Types of property 'toString' are incompatible.
              Type '() => string' is not assignable to type 'ExactObject<() => string, never>'.
                Type '() => string' is not assignable to type 'Record<string | number | symbol, never>'.
                  Index signature for type 'string' is missing in type '() => string'.
```

Error at `testArg(t8)`:
```
error TS2345: Argument of type '{ a: string; b: Date | null; }' is not assignable to parameter of type 'ExactObject<A, { a: string; b: Date | null; }>'.
  Type '{ a: string; b: Date | null; }' is not assignable to type '{ a: string; b?: ExactObject<Date, Date | null> | null | undefined; }'.
    Types of property 'b' are incompatible.
      Type 'Date | null' is not assignable to type 'ExactObject<Date, Date | null> | null | undefined'.
        Type 'Date' is not assignable to type 'ExactObject<Date, Date | null>'.
          Type 'Date' is not assignable to type '{ toString: ExactObject<() => string, never>; toDateString: ExactObject<() => string, never>; toTimeString: ExactObject<() => string, never>; ... 41 more ...; [Symbol.toPrimitive]: ExactObject<...>; }'.
            Types of property 'toString' are incompatible.
              Type '() => string' is not assignable to type 'ExactObject<() => string, never>'.
```

<!-- POLAR PLEDGE BADGE START -->
## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/896">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/896/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/sindresorhus/type-fest/issues/896/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

--- a/source/exact.d.ts
+++ b/source/exact.d.ts
@@ -1,7 +1,7 @@
 import type {ArrayElement, ObjectValue} from './internal';
-import type {Opaque, TagContainer} from './opaque';
 import type {IsEqual} from './is-equal';
 import type {KeysOfUnion} from './keys-of-union';
+import type {JsonObject} from './basic';
 
 /**
 Create a type from `ParameterType` and `InputType` and change keys exclusive to `InputType` to `never`.
@@ -57,7 +57,6 @@ export type Exact<ParameterType, InputType> =
 		: ParameterType extends unknown[] ? Array<Exact<ArrayElement<ParameterType>, ArrayElement<InputType>>>
 			// In TypeScript, Array is a subtype of ReadonlyArray, so always test Array before ReadonlyArray.
 			: ParameterType extends readonly unknown[] ? ReadonlyArray<Exact<ArrayElement<ParameterType>, ArrayElement<InputType>>>
-				// Leave tagged types as-is. We could try to make the untagged part Exact, and just leave the tag as-is, but that seems to create instanitation excessively deep errors.
-				: ParameterType extends TagContainer<unknown> ? ParameterType
-					: ParameterType extends object ? ExactObject<ParameterType, InputType>
-						: ParameterType;
+				// Only apply Exact for pure object types. For types from a class, leave it unchanged to TypeScript to handle.
+				: ParameterType extends JsonObject ? ExactObject<ParameterType, InputType>
+					: ParameterType;

--- a/test-d/exact.ts
+++ b/test-d/exact.ts
@@ -450,3 +450,20 @@ import type {Exact, Opaque} from '../index';
 		},
 	});
 }
+
+// Spec - special test case for Date type + union
+// @see https://github.com/sindresorhus/type-fest/issues/896
+{
+	type A = {
+		a: string;
+		b?: Date | null;
+	};
+
+	const function_ = <T extends Exact<A, T>>(arguments_: T) => arguments_;
+
+	function_({a: 'a'});
+	function_({a: 'a', b: new Date()});
+	function_({a: 'a', b: new Date() as Date | null});
+	// @ts-expect-error
+	function_({a: 'a', b: 1});
+}

